moreno20241120

MorenoMarrara opened the pull request.

Commit:
implementation of server logic

Commit:
implemented Server.createLobby()

albiosz commented on client/pom.xml:
```diff
@@ -9,8 +9,8 @@
     <properties>
-        <maven.compiler.source>21</maven.compiler.source>
-        <maven.compiler.target>21</maven.compiler.target>
+        <maven.compiler.source>17</maven.compiler.source>
```
Why did you change the version of java?

albiosz commented on shared/src/main/java/alcatraz/shared/Lobby.java:
```diff
@@ -6,42 +6,51 @@
 
-    public boolean addPlayer(Player player) {
-        if (this.players.size() < 4) {
+    public void addPlayer(Player player) throws RemoteException {
```
The error could be more specific, the RemoteException is thrown when something is wrong with RMI connection, so it can be even misleading.
I think in this case we should create a new error e.g. LobbyIsFull or something like that.

MorenoMarrara replied:
see LobbyFullException

albiosz commented on shared/src/main/java/alcatraz/shared/LobbyKey.java:
```diff
@@ -0,0 +1,12 @@
+package alcatraz.shared;
+
+public class LobbyKey {
+    public final long id;
```
What for is the id of the key? Should not the secret already be unique, thus identifies the key 🤔

MorenoMarrara replied:
was renamed to lobbyId to make it clearer

albiosz commented on shared/pom.xml:
```diff
@@ -9,8 +9,8 @@
     <properties>
-        <maven.compiler.source>21</maven.compiler.source>
-        <maven.compiler.target>21</maven.compiler.target>
+        <maven.compiler.source>17</maven.compiler.source>
```
why?

albiosz commented on server/pom.xml:
```diff
@@ -9,8 +9,8 @@
     <properties>
-        <maven.compiler.source>21</maven.compiler.source>
-        <maven.compiler.target>21</maven.compiler.target>
+        <maven.compiler.source>17</maven.compiler.source>
```
why?

albiosz commented on server/src/main/java/alcatraz/server/state/LobbyManager.java:
```diff
@@ -1,34 +1,78 @@
+    *           When an owner creates a new lobby and they have been inside a lobby before, they will be relocated to
+    *               the newly created lobby.
+     */
+    public LobbyKey createLobby(Player owner) throws RemoteException {
```
the exception should be more specific

MorenoMarrara replied:
"throws RemoteException" was removed as method it shouldn't throw anyways

albiosz commented on server/src/main/java/alcatraz/server/state/LobbyManager.java:
```diff
@@ -1,34 +1,78 @@
-        // TODO: it needs to be checked of there is still space for another player, maximal 4 players
-        return lobbies.get(lobbyId).addPlayer(player); // vielleicht kann das ein void sein mit exception?
+
+    public void addPlayerToLobby(long lobbyId, Player player) throws RemoteException {
```
the exception could be more precise

MorenoMarrara replied:
I guess it can't, as removePlayerFromLobby() throws LockedLobbyException and addPlayer() throws RemoteException

albiosz commented on server/src/main/java/alcatraz/server/state/LobbyManager.java:
```diff
@@ -1,34 +1,78 @@
+        lobbies.get(lobbyId).addPlayer(player);
+    }
+
+    public void removePlayerFromLobby(String playerName) throws RemoteException {
```
the exception could be more precise

albiosz commented on server/src/main/java/alcatraz/server/state/LobbyManager.java:
```diff
@@ -1,34 +1,78 @@
+            throw new RemoteException();        //MM20241122: find better Exception! "MAXSIZE reached"
+        }
+
+        while (lobbies.containsKey(lobbyIdCounter)) {
```
Is it really more efficient to look for a first free id, instead of generating random lobby id? 🤔

MorenoMarrara replied:
it was the simplest mechanism, that I could think of and that also works

albiosz commented on server/src/main/java/alcatraz/server/Server.java:
```diff
@@ -22,68 +19,69 @@ public Server(ReplicationInterface replication) throws RemoteException {
+            System.out.println(clientName + " registered");
+            this.replication.replicatePrimaryState();
             // TODO: there should be an early return instead of else
+            //MM20241121: why?
```
It would demand less cognitive effort.
```java
if (!isNotPrimary) {
  forward();
  return;
}
// the rest function logic
```
Now we have to find the else every time that is 13 lines under the `if`, and then have to look even further if it is the last statement in the function. With early return you can see directly all the actions that happen if the server is not primary.

MorenoMarrara replied:
totally agree! didn't know what you meant by early exit!

albiosz commented on server/src/main/java/alcatraz/server/Server.java:
```diff
@@ -22,68 +19,69 @@ public Server(ReplicationInterface replication) throws RemoteException {
+            LobbyKey key = this.state.lobbyManager.createLobby(owner);
+            System.out.println("Lobby created with ID: " + key.id + " by Player: " + owner.getClientName());
+
+            this.state.lobbyManager.addPlayerToLobby(key.id, owner);
```
we also will need a mutex here, won't we? 🤔

albiosz commented on server/src/main/java/alcatraz/server/Server.java:
```diff
@@ -22,68 +19,69 @@ public Server(ReplicationInterface replication) throws RemoteException {
     @Override
-    public boolean joinLobby(String clientName, Long lobbyId) throws RemoteException {
+    public void joinLobby(Player client, Long lobbyId) throws RemoteException {
         if (this.replication.isPrimary()) {
```
I would decrease the cyclomatic complexity of the function. From my perspective it is no not really readable now.

I think something like this would be more readable and will grow better in case there is more logic to add.
```java
if (!isPrimary) {
  // forwarding
  return
}
if (!doesNotContainKey) {
  // the exception
  return
}
// the rest
```

albiosz commented on client/src/main/java/alcatraz/client/Main.java:
```diff
@@ -69,8 +70,8 @@ public static void main(String[] args) {
                         System.out.println("Creating Lobby...");
-                        Lobby lobby = server.createLobby(clientName);
-                        System.out.println("Lobby created with id: " + lobby.getId());
+                        LockedLobby lobby = server.createLobby(clientName);
```
I got an error a compilation error because the `LockeLobby` was not found

albiosz reviewed (changes requested):
The code does not compile, please correct the errors

albiosz committed:
add implementation of server logic

albiosz commented on shared/src/main/java/alcatraz/shared/DuplicateNameException.java:
```diff
@@ -0,0 +1,10 @@
+package alcatraz.shared;
```
the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

MorenoMarrara replied:
totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

albiosz commented on shared/src/main/java/alcatraz/shared/LobbyFullException.java:
```diff
@@ -0,0 +1,9 @@
+package alcatraz.shared;
```
the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

MorenoMarrara replied:
totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

albiosz commented on shared/src/main/java/alcatraz/shared/LobbyLockedException.java:
```diff
@@ -0,0 +1,9 @@
+package alcatraz.shared;
```
move the exception to the separate sub package

MorenoMarrara replied:
totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

albiosz commented on shared/src/main/java/alcatraz/shared/NotEnoughPlayersException.java:
```diff
@@ -0,0 +1,11 @@
+package alcatraz.shared;
```
the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

MorenoMarrara replied:
totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

albiosz commented on shared/src/main/java/alcatraz/shared/PlayerNotRegisteredException.java:
```diff
@@ -0,0 +1,10 @@
+package alcatraz.shared;
```
the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

MorenoMarrara replied:
totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

albiosz commented on shared/src/main/java/alcatraz/shared/TooManyLobbiesException.java:
```diff
@@ -0,0 +1,10 @@
+package alcatraz.shared;
```
the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

MorenoMarrara replied:
totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

albiosz commented on server/src/main/java/alcatraz/server/Server.java:
```diff
@@ -2,146 +2,150 @@
-            System.out.println(playerName + " registered");
-            return true;
-            // TODO: there should be an early return instead of else
+            //MM20241121: use a mutex at this stage!
```
outdated comment

albiosz commented on server/src/main/java/alcatraz/server/Server.java:
```diff
@@ -2,146 +2,150 @@
-            return true;
-            // TODO: there should be an early return instead of else
+            //MM20241121: use a mutex at this stage!
+            this.state.players.put(clientName, new Player(client, clientName, "ip_address", "port"));
```
ip_address and port should come from the client.
It would be nice to add TODO at this point and implement the logic in another PR,
or implement it directly, by changing the signature of the function to include `ipAddress` and `port` too

MorenoMarrara replied:
I agree, but does the server have to authenticate the ip_address or does it simply believe what it's told?

albiosz replied:
I'd say that in the first version the server should trust, that the client passed a correct IP and in further versions we can improve the process, if there is need (like using the function `client.isAvailable()`)

albiosz replied:
I changed the this thread to a separate issue

albiosz commented on server/src/main/java/alcatraz/server/Server.java:
```diff
@@ -2,146 +2,150 @@
-            this.state.lobbyManager.createLobby(lobbyId, lobby);
-            System.out.println("Lobby created with ID: " + lobbyId + " by Player: " + clientName);
+            fairLock.lock();
+            //MM20241121: This section needs a mutex
```
outdated comment

albiosz commented on server/src/main/java/alcatraz/server/Server.java:
```diff
@@ -2,146 +2,150 @@
+            Lobby gameLobby = this.state.lobbyManager.getLobbyById(lobbyId);
+            if (gameLobby.checkSecret(secret)) {
+                gameLobby.setUnavailable();
+                gameLobby.getPlayers().forEach((key, value) -> {
```
change the names of the parameters (key, value) to the names that describe what the variables contain e.g. (playerName, player)

albiosz reviewed (changes requested):
- I would refactor all the methods in the Server.java class, so that they early return where it is possible, instead of `else` statements
e.g.
```java
if (!primary) {
  replicate();
  return;
}

if (!checkSecret()) {
    throw new CheckSecretException();
}
```
It should considerable improve the quality (lower cyclomatic complexity) of the code.

Commit:
Merge remote-tracking branch 'origin/moreno20241120' into moreno20241120

# Conflicts:
#	client/src/main/java/alcatraz/client/Main.java
#	server/src/main/java/alcatraz/server/Server.java
#	server/src/main/java/alcatraz/server/state/LobbyManager.java
#	shared/src/main/java/alcatraz/shared/Lobby.java
#	shared/src/main/java/alcatraz/shared/LobbyKey.java
#	shared/src/main/java/alcatraz/shared/ServerInterface.java

Commit:
MM20241129

Replied to Albert's comments and made changes where I
dared making them.

20241128

corrected building errors

20241127
implemented according to today's discussion
cannot think anymore but wanted to share nonetheless

newly implemented
Lobby
 - improved Lobby::addPlayer()
  . added sanity checks before a player can be added
 - improved Lobby::removePlayer()
  . added check for availability: project description
    says that "no player may leave the Lobby after
    game start"
 - improved Lobby availability management for game
   start and adding players.

LobbyKey
 - renamed lobbyId member for better readability

LobbyManager
 - implemented LobbyManager::getAllLobbies() for
   replication
 - implemented LobbyManager::getAvailableLobbies() for
   ClientInterface
 - implemented LobbyManager::getLobbyIdByClientName()
   -> maybe redundant
 - improved LobbyManager::addPlayerToLobby() so that
   player will be removed from a lobby if it was in
   another before and then added to the newly provided
 - improved LobbyManager::removePlayerFromLobby() so
   that we correctly track player and lobby. Project
   description says that "each player can only be in
   one lobby at a time"
 - implemented LobbyManager::isLobbyReadyForGameStart
   for better readability

Server
 - implemented Server::isPlayerRegistered() for checks
 - improved Server::joinLobby() with check for
   registration and improved readability
 - implemented Server::leaveLobby()
 - improved Server::initializeGameStart() with mutex
   and lobby readyness before game start

ServerInterface
 - aligned and extended interface with needs by Server
   class

useCasesImplemented.txt
 - collection of implementation details and TODOS

Commit:
MM20241129

Replied to Albert's comments and made changes where I
dared making them.

20241128

corrected building errors

20241127
implemented according to today's discussion
cannot think anymore but wanted to share nonetheless

newly implemented
Lobby
 - improved Lobby::addPlayer()
  . added sanity checks before a player can be added
 - improved Lobby::removePlayer()
  . added check for availability: project description
    says that "no player may leave the Lobby after
    game start"
 - improved Lobby availability management for game
   start and adding players.

LobbyKey
 - renamed lobbyId member for better readability

LobbyManager
 - implemented LobbyManager::getAllLobbies() for
   replication
 - implemented LobbyManager::getAvailableLobbies() for
   ClientInterface
 - implemented LobbyManager::getLobbyIdByClientName()
   -> maybe redundant
 - improved LobbyManager::addPlayerToLobby() so that
   player will be removed from a lobby if it was in
   another before and then added to the newly provided
 - improved LobbyManager::removePlayerFromLobby() so
   that we correctly track player and lobby. Project
   description says that "each player can only be in
   one lobby at a time"
 - implemented LobbyManager::isLobbyReadyForGameStart
   for better readability

Server
 - implemented Server::isPlayerRegistered() for checks
 - improved Server::joinLobby() with check for
   registration and improved readability
 - implemented Server::leaveLobby()
 - improved Server::initializeGameStart() with mutex
   and lobby readyness before game start

ServerInterface
 - aligned and extended interface with needs by Server
   class

useCasesImplemented.txt
 - collection of implementation details and TODOS
